File: day3/solution3.1.ts
import * as fs from "fs";

function sumValidMulResults(inputFile: string): number {
    // Read the input file
    const data = fs.readFileSync(inputFile, "utf8");

    // Regular expression to match valid mul(X,Y) instructions
    const pattern = /mul\((\d+),(\d+)\)/g;

    // Find all matches
    let match;
    let sum = 0;

    while ((match = pattern.exec(data)) !== null) {
        const x = parseInt(match[1], 10);
        const y = parseInt(match[2], 10);
        sum += x * y;
    }

    return sum;
}

// Example usage
const inputFile = "./input.txt";
console.log("Sum of valid mul results:", sumValidMulResults(inputFile));

File: day3/solution3.1.py
import re

def sum_valid_mul_results(input_file):
    # Read the input file
    with open(input_file, "r") as file:
        data = file.read()
    
    # Regular expression to match valid mul(X,Y) instructions
    pattern = r"mul\((\d+),(\d+)\)"
    
    # Find all matches
    matches = re.findall(pattern, data)
    
    # Compute the sum of all valid multiplications
    result = sum(int(x) * int(y) for x, y in matches)
    
    return result

if __name__ == "__main__":
    input_file = "./input.txt"
    print("Sum of valid mul results:", sum_valid_mul_results(input_file))